[libc] Improve qsort

Voultapher opened the pull request:
This PR improves the libc `qsort` and `qsort_r` implementation in a variety of ways:

- Vastly improved performance, 2+x improvement in many scenarios, see below.

- Improved usage safety, the current implementation allows for trivial out-of-bounds read and write UB if the comparison function does not implement a strict weak ordering. I've written about this in more depth [here](https://github.com/Voultapher/sort-research-rs/blob/main/writeup/sort_safety/text.md#ord-safety). For reference glibc qsort considers this a [safety critical bug](https://www.openwall.com/lists/oss-security/2024/01/30/7) and does bounds checking. The current implementation contains code like this `while ((compare_i = array.elem_compare(i, pivot)) < 0) ++i;` which goes out-of-bounds if the comparison function does not implement a strict weak ordering. The new implementation either does bounds checking or loop bounds independent of the comparison result.

- Better worst case algorithmic complexity. The current implementation is a classical quicksort implementation with a worst case expected time to sort the data of O(N^2). The new implementation improves that to O(N x log(N)) by limiting the recursion depth and switching to heapsort if a limit is hit, this idea comes from introsort. In addition the new implementation achieves O(N x log(K)) where K is the number of distinct elements by using equal element filtering as pioneered by [pdqsort](https://github.com/orlp/pdqsort).

There is one downside, binary-size. The current implementation is very basic and uses dynamic dispatch to re-use the same implementation for `qsort` and `qsort_r`. In practice with default `-O3` settings on x86-64 the binary-size `qsort` and `qsort_r` contribute in total to libc goes from 1kB to 18kB, this is a net 17kB increase. It's possible to reduce binary-size further but doing so comes at the price of significant performance reduction. The new implementation - idisort (ipnsort derived introsort sort) - makes rather conservative choices, with a higher binary-size budget things like efficient full ascending and descending handling or more size specializations are possible.

For some background, together with @orlp I worked on improving the Rust standard library sort implementations which was merged earlier this year and is part of stable Rust for several month now. The ideas in idisort are largely derived from the learnings gained by developing [ipnsort](https://github.com/Voultapher/sort-research-rs/blob/main/writeup/ipnsort_introduction/text.md). The major performance difference comes from having an opaque comparison function as forced by the `qsort` interface, which limits the kind of optimizations that can be done in code an by the compiler.

### Benchmark setup

```
Linux 6.11
clang version 18.1.8
rustc 1.85.0-nightly (28fc2ba71 2024-11-24)
AMD Ryzen 9 5900X 12-Core Processor (Zen 3 micro-architecture)
CPU boost enabled.
```

### Benchmark results

The used methodology is explained in detail [here](https://github.com/Voultapher/sort-research-rs/blob/main/writeup/ipnsort_introduction/text.md#efficient).

The *current* LLVM libc qsort implementation is referenced as `c_llvm_libc_unstable`, the new implementation is referenced as `c_idisort_unstable`. There are other implementations included for a comparison of the state of the art.

<img src="https://github.com/user-attachments/assets/13592b7c-edd1-4e4a-998a-6e1da9593d69" width=700 />

This graph compares the performance of two implementations against each other, values above zero imply speedups for the new implementation and values below zero imply slowdowns. Each dot represents one input length for a specific pattern combination.

<img src="https://github.com/user-attachments/assets/2d64bd19-bf55-4de8-aab3-6848ff2658da" width=960 />

Expand this for additional benchmark results with other types:

<details>
<img src="https://github.com/user-attachments/assets/b3f0ea45-64ef-4601-a376-f14d6f199595" width=960 />

<img src="https://github.com/user-attachments/assets/a618dca0-2304-4fcd-9923-aee108d34b2c" width=960 />

`random_p5` at an input length of 22k peaks at a 558x performance improvement.
Keep in mind this is an artifical test. But generally large types see quite bad performance in the current implementation.

<img src="https://github.com/user-attachments/assets/c99d5795-c77b-49d7-903c-7ff71b6746d8" width=960 />

<img src="https://github.com/user-attachments/assets/cf3ef6fd-656a-4199-b567-d70e4136f074" width=960 />
</details>

This PR is best reviewed by individual commits, they represent logical increments that are all individually buildable and correct. The commit messages explain the reasoning behind changes in more depth.

This is my first time contributing to the LLVM libc so please point out structural mistakes.

Voultapher committed:
Use templated is_less instead of runtime comparator

This changes serves both immediate purposes and lays the groundwork for further
changes. Previously the qsort implementation used a comparator construct that
dispatched the comparison function through a tagged union, this incurs an
additional branch on each comparison to check the tag `comp_type`. There are
also structural changes to the quick_sort implementation to lay the groundwork
for future changes to the partition and equal element handling logic. The
comparator also did an equal idx comparison for each comparison. While this may
sound like a nice optimization for expensive to compare types, a well
constructed implementation will never compare an element to itself. In effect
this change reduces the required comparison for integer sorting from 4 to 2 per
comparison operation in the implementation.

Performance impact, the tested types i32, u64, string and f128 improve by ~1.5x,
and 1k by ~2.4x for the random pattern. Equal element patterns see an
algorithmic degradation, but this will be addressed by a future change.

Nice to have, the `[[clang::no_sanitize("function")]]` is no longer needed.

Voultapher committed:
Use higher quality pivot selection

This commit ports the glidesort recursive median pivot selection. Without a
dedicated small-sort this pivot selection is worse for cheap to compare types,
but it is a precursor to optimally leverage pdqsort style equal element
filtering.

Voultapher committed:
Use pdqsort style equal element filtering

This commits the technique introduced by Orson Peters in pdqsort to efficiently
handle inputs with repeated elements for example as found in zipfian or low
cardinality distributions. The random_z1 pattern sees a ~3.5x improvement in
performance for input length 10k across the different types. This change implies
O(N x log(K)) for inputs with K distinct elements.

Voultapher committed:
Use heapsort fallback to guarantee O(N x log(N)) worst case performance

The combination of a high quality pivot selection function with equal element
filtering as added in the previous commit makes for quicksort that in practice
hardly ever hits quadratic run-time scaling, but malicious inputs and accidents
are possible. By limiting the recursion depth and switching to heapsort -
introsort the second i in idisort - we can guarantee a worst case expected time
to sort the data of O(N x log(N)).

Voultapher committed:
Use fixed size array specialization

By itself this optimization only nets a ~1.2x improvement for types like i32,
u64 and f128, but it unlocks ~2x improvements for partitioning. The cost is a
modest increase in binary-size of libc, `unstable_sort` is only instantiated
twice, which raises the total from 2 to 8 instantiations in the library. But
since the fixed size instantiations have much cheaper swap routines they only
add ~1.5k each vs the ~4.5k generic instantiation. In effect binary size for
qsort and qsort_r goes from ~9k to ~18k.

Voultapher committed:
Use branchless lomuto partition when elem size is known at compile-time

Uses a simplified version of the branchless partition scheme used in ipnsort.
This achieves a ~2.5x perf improvement for i32, u64 and f128 for random
patterns. This is mainly achieved by avoiding branch misprediction penalties.

Voultapher committed:
Use more generic dynamic element size swap impl

The new impl is better for types < 100 bytes and worse for very large types like
1k, but still not bad, plus it has smaller binary-size footprint and avoids a
surprising performance cliff past `STACK_ARRAY_SIZE` as well as heuristics.

Also remove inline "hints" it's the wrong way to do it and the situation is
simple enough that the compiler should now best and it's the wrong way to hint
it anyway.

Voultapher committed:
Avoid unnecessary pivot swap for common recursion case

The pivot selection will pick 0 as the pivot index if the array length is < 8,
which is a common recursion leaf case. This isn't a big perf win more on the
order 1-2% but's it's trivial and more or less free.

Voultapher commented on libc/src/stdlib/qsort_data.h:
This part feels a little sketchy to me, could we run into alignment issues, the only kind of usage is loading and storing the data, it's not interpreted in some semantic sense, but still I could imagine some hardware that can't handle for example a `char*` addressed as `uint32_t`. Could we maybe check the `array_base` pointer alignment IIRC alignment implies the number of zeros at the end or so. I'm certainly no expert on memory alignment semantics so I'd appreciate help with this part.

Voultapher replied:
Doesn't that code also run into strict aliasing issues?

Voultapher replied:
I've created https://github.com/llvm/llvm-project/issues/120709 to follow up on that question.

michaelrj-google commented on libc/src/stdlib/qsort_util.h:
```diff
@@ -27,11 +27,46 @@
 namespace LIBC_NAMESPACE_DECL {
 namespace internal {
 
+template <typename A, typename F> void sort_inst(A &array, const F &is_less) {
```
Naming the comparison `is_less` is a bit confusing since it's not a binary comparison. The comparison function returns a positive, negative, or zero integer if the first element is greater than, less than, or equal to.

Voultapher replied:
Uhh, it is a binary comparison .. the name is the same used in the code it is ported from. Take a look at how it is used in the code.

michaelrj-google replied:
The name `is_less` implies it returns a boolean, e.g. "a is less than b". Using a name like `compare` makes it more clear this a comparison function that returns a direction, similar to strcmp. A more clear name makes it easier for everyone to understand the code, you included. See my comment in `quick_sort.h`.

orlp replied:
@michaelrj-google I do believe it returns a boolean, see [`libc/src/stdlib/qsort.cpp`](https://github.com/llvm/llvm-project/pull/120450/files/efb98579a7f56c721402d526086fa24a9cbc3110#diff-26ff238bd5dfa4e151c74fb30555d8b07621bbf887766762dbf4774f1c9f5c2f) where `internal::quicksort` is instantiated with `compare(a, b) < 0`.

michaelrj-google commented on libc/test/src/stdlib/SortingTest.h:
```diff
@@ -66,45 +68,36 @@ class SortingTest : public LIBC_NAMESPACE::testing::Test {
   }
 
-  void test_reversed_sorted_array(SortingRoutine sort_func) {
+  void test_reversed_sorted_array(SortFn sort_fn) {
```
if you'd like to rename these variables that might fit better as a followup patch since it's not really related to this patch.

Voultapher replied:
Fair point, I had originally removed that parameter and then later added it again, should have gone with the same name. Will change.

michaelrj-google commented on libc/test/src/stdlib/quick_sort_test.cpp:
```diff
@@ -1,16 +0,0 @@
-//===-- Unittests for quick sort ------------------------------------------===//
```
this is important to keep for when heap sort is configured as the default sorting algorithm. We want to make sure we're testing all of the options.

Voultapher replied:
Genuine question, can you describe a scenario where not having this test would miss an actual issue? Presumably it's not always going to be configured as the default right?

michaelrj-google replied:
We want to test every sorting implementation regardless of what's the default. This file ensures we're testing the quick sort implementation, just like the `heap_sort_test` file ensures we're testing the heap sort implementation. 

A scenario where not having this test would miss an issue would be on a platform where `LIBC_QSORT_IMPL == LIBC_QSORT_HEAP_SORT` (e.g. the baremetal targets). At that point both the `qsort_test` and `heap_sort_test` are checking heapsort, with no testing for quick sort. Someone developing for that platform might make changes to the shared sorting code, but miss any breakages in quick sort.

Voultapher replied:
I had assumed both defines are always tested in the CI, by doing testing it this way we test the impl but not if actual user interface of setting a compile time define. I'll add the quick_sort test back in.

Voultapher replied:
Since we are not testing the public interface anyway, I've changed it to test quick_sort and heap_sort explicitly with as little code repeated as sensible in the tests.

michaelrj-google reviewed (commented):
Thanks for the patch! Improving the qsort algorithm would be very appreciated. I've left some comments, and I hope we can land this soon.

Voultapher commented on libc/src/stdlib/qsort_data.h:
Good to know, since a later commit removes this code I'll mark this as resolved.

Voultapher commented on libc/src/stdlib/qsort_data.h:
For what it's worth, since this implementation doesn't use temporary copies of the data - with the exception of the swap routine which doesn't call `is_less` - there aren't any problems and it ought to provide [intuitive exception safety](https://github.com/Voultapher/sort-research-rs/blob/main/writeup/sort_safety/text.md#exception-safety). I'll remove the noexcepts.

Voultapher replied:
Done.

Voultapher commented on libc/src/stdlib/qsort_pivot.h:
```diff
@@ -0,0 +1,92 @@
+    // If x=y=1 then a < b, c. In this case we want to return min(b, c).
+    // By toggling the outcome of b < c using XOR x we get this behavior.
+    const bool z = is_less(b_ptr, c_ptr);
+    if (z ^ x) {
+      return c;
+    } else {
+      return b;
+    }
```
Is this a style-guide thing? Personally I much prefer to have those for sound sleep.

michaelrj-google replied:
Yeah, the LLVM style guide recommends skipping them in places where it's obvious: https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

nickdesaulniers replied:
Could also just use a ternary:

```c
return z ^ x ? c : b;
```
is pretty concise.

Voultapher commented on libc/src/stdlib/qsort_util.h:
```diff
@@ -27,11 +27,46 @@
+  uint8_t *array_base = reinterpret_cast<uint8_t *>(array);
+
+  switch (elem_size) {
+  case 4: {
+    auto arr_fixed_size = internal::ArrayFixedSize<4>(array_base, array_len);
+    sort_inst(arr_fixed_size, is_less);
+    return;
+  }
+  case 8: {
+    auto arr_fixed_size = internal::ArrayFixedSize<8>(array_base, array_len);
+    sort_inst(arr_fixed_size, is_less);
+    return;
+  }
+  case 16: {
+    auto arr_fixed_size = internal::ArrayFixedSize<16>(array_base, array_len);
+    sort_inst(arr_fixed_size, is_less);
+    return;
+  }
```
Ok, I'll add it, any idea what the `#ifdef` should be? Is there an existing one?

Voultapher replied:
If all you care about is binary-size over everything we should probably configure it to use heap_sort only, more or less same algorithmic guarantees, but slower but also much smaller. It's what the Rust stdlib [does](https://github.com/rust-lang/rust/pull/129587).

Voultapher replied:
I see, since it only helps in terms of binary-size and no longer in terms of worst case run-time complexity, I suggest a better name for that ifdef. Something along the lines of `#ifdef OPTIMIZE_FOR_SIZE`.

Voultapher commented on libc/src/stdlib/quick_sort.h:
```diff
@@ -9,84 +9,172 @@
 
-#include "src/__support/macros/attributes.h"
+#include "src/__support/CPP/cstddef.h"
+#include "src/__support/big_int.h"
```
IIRC that's where `cpp::bit_width` comes from.

michaelrj-google commented on libc/src/stdlib/quick_sort.h:
```diff
@@ -9,84 +9,172 @@
-    } else if (j == pivot_index) {
-      pivot = array.get(i);
-      pivot_index = i;
+    while (left < right && is_less(array.get(left), pivot))
+      ++left;
+
+    while (true) {
+      --right;
+      if (left >= right || is_less(array.get(right), pivot)) {
+        break;
+      }
```
these loops are incorrect, since the comparison function currently named `is_less` will only return `0` when the two values are equal. Both will iterate until the values at `left` and `right` are equal to `pivot`.

Voultapher replied:
`is_less` returns `bool` and does exactly what the name suggests it does. Note the previous Hoare partition implementation had a common bug where it needlessly repeated a comparison at the end, something nearly everyone gets wrong, but it's not too impactful.

michaelrj-google commented on libc/test/src/stdlib/SortingTest.h:
```diff
@@ -28,16 +27,19 @@ class SortingTest : public LIBC_NAMESPACE::testing::Test {
   }
 
-  void test_sorted_array(SortingRoutine sort_func) {
+  static void int_sort(SortFn sort_fn, int *array, size_t array_len) {
+    sort_fn(reinterpret_cast<void *>(array), array_len, sizeof(int),
+            int_compare);
+  }
```
if we're going to have different partitioning algorithms based on the size of the input values, we need to make sure to test both of those partitioning algorithms. I'd suggest adding a uint128 test.

Voultapher replied:
Good idea, I've added a new test that covers a range of element sizes, `uint128` wouldn't help because it also uses the Lomuto partition.

SchrodingerZhu commented on libc/src/stdlib/qsort_data.h:
We prefer to mark all header functions with `LIBC_INLINE` if applicable. Such macro allows us to apply other attributes (such as always_inline or `noinline` for testing purposes) to all inline-able functions.

SchrodingerZhu commented on libc/src/stdlib/qsort_util.h:
```diff
@@ -27,11 +27,46 @@
+void unstable_sort(void *array, size_t array_len, size_t elem_size,
+                   const F &is_less) {
+  if (array == nullptr || array_len == 0 || elem_size == 0) {
+    return;
```
Please remove the extra quotes following https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

SchrodingerZhu commented on libc/src/stdlib/qsort_util.h:
```diff
@@ -27,11 +27,46 @@
+    return;
+  }
+
+  uint8_t *array_base = reinterpret_cast<uint8_t *>(array);
```
`static_cast` should be more appropriate here.

See item `(g` at https://en.cppreference.com/w/cpp/language/static_cast

Voultapher replied:
No longer relevant.

SchrodingerZhu commented on libc/src/stdlib/qsort_data.h:
it should work without the explicit casting.

SchrodingerZhu commented on libc/src/stdlib/qsort_data.h:
In rust, `*mut u8` is used as the low-level byte pointers. The `cpp::byte*` is more preferable. Please see https://en.cppreference.com/w/cpp/types/byte.

Voultapher replied:
Good point, will do.

SchrodingerZhu commented on libc/src/stdlib/qsort_data.h:
You can use https://github.com/llvm/llvm-project/blob/f334db92be168876b618db72dc93078ce23ffa89/libc/src/__support/CPP/array.h to avoid the possible alignment issue. Then, it is up to the compiler to decide whether the microarch supports reading unaligned word cells.
```suggestion
using block_t = cpp::array<cpp::byte, 4>;
```

SchrodingerZhu commented on libc/src/stdlib/qsort_data.h:
@Lancern is pointer laundering needed here?

Lancern replied:
Pointer laundering does not help here. Beside the potential alignment issues, it inevitably violates strict aliasing. Do we have strict aliasing disabled in libc?

SchrodingerZhu replied:
I think a lot of libc code is compiled with -O3 without `-fno-strict-aliasing`

SchrodingerZhu replied:
```c++
struct Block {
  std::byte data[4];
  static Block load(const std::byte* src) {
     block_t buf;
     #if __has_builtin(__builtin_memcpy_inline)
          __builtin_memcpy_inline(buf.data, src, 4);
     #else
          __builtin_memcpy(buf.data, src, 4);
     #endif
  } 
  /* store similarly */
};
```
Then you can use `Block::load(...)/Block::store(...)` to do the operation.
The compiler is smart enough to optimize the code into what you are expecting.

Voultapher replied:
Good idea, I'll apply it although I'd like to go with a less abstraction heavy approach that is hopefully easier to understand and maintain. What is the difference between using `__builtin_memcpy_inline` and `__builtin_memcpy` here, because in both cases the compiler ought to replace it with inline operations if those are possible on the platform since the size is a constant?

SchrodingerZhu replied:
__builtin_memcpy_inline has the guarantee that no external function call will be emitted but it is clang specific. 
>  both cases the compiler ought to replace it with inline operations
Yes, In general, `gcc/clang` treat memcpy as internal functions. For fix sized memcpy, it is almost certain that no call is needed.

SchrodingerZhu reviewed (commented):
Thank you very much for the patch. I added some comments during my review. I haven't finished reviewing later parts but the code and results look very promising to me.

SchrodingerZhu commented on libc/src/stdlib/qsort_r.cpp:
```diff
@@ -19,13 +19,12 @@ LLVM_LIBC_FUNCTION(void, qsort_r,
-  internal::sort(arr);
+  internal::unstable_sort(
+      array, array_size, elem_size,
+      [compare, arg](const void *a, const void *b) noexcept -> bool {
```
@michaelrj-google I think the `is_less` function @Voultapher passed in is actually a wrapper of the raw integer function. So, calling it `is_less` should be good.

Voultapher commented on libc/src/stdlib/qsort_pivot.h:
```diff
@@ -0,0 +1,92 @@
+#ifndef LLVM_LIBC_SRC_STDLIB_QSORT_PIVOT_H
+#define LLVM_LIBC_SRC_STDLIB_QSORT_PIVOT_H
+
+#include "src/stdlib/qsort_pivot.h"
```
Good catch, that's nonsense.

Voultapher committed:
Apply review comments

Voultapher commented:
I've combined all the style and minor changes into a first commit. I'll tackle the incorrect `swap` impl next.

Voultapher committed:
Fix strict aliasing violation in `swap` function

There might be platform where addressing the user provided array of unknown
alignment as `uint32_t` would be incorrect. Generally the previous code violates
strict aliasing and thus invokes UB. By using the "magic" properties of `memcpy`
we can sidestep the issue by letting it do the load and store for us. In
practice this means on platforms where addressing the memory as 4 byte regions
is safe and efficient to do - which is most widely used platforms - the compiler
can insert the appropriate `mov` instructions inline and if a platform does not
support this operation it will likely insert a call to the `memcpy` function.

Voultapher commented:
I've also fixed the strict aliasing violation, this should be everything that was raised in the review so far, please take another look.

SchrodingerZhu commented on libc/src/stdlib/qsort_pivot.h:
```diff
@@ -0,0 +1,89 @@
+  const size_t len = array.len();
+
+  if (len < 8) {
+    return 0;
```
This brace can also be removed.

SchrodingerZhu commented on libc/src/stdlib/qsort_pivot.h:
```diff
@@ -0,0 +1,89 @@
+  const size_t c = len_div_8 * 7; // [7*floor(n/8), 8*floor(n/8))
+
+  if (len < PSEUDO_MEDIAN_REC_THRESHOLD) {
+    return median3(array, a, b, c, is_less);
```
ditto

SchrodingerZhu commented on libc/src/stdlib/heap_sort.h:
```diff
@@ -18,11 +18,12 @@ namespace internal {
-LIBC_INLINE void heap_sort(const Array &array) {
-  size_t end = array.size();
+template <typename A, typename F>
+void heap_sort(const A &array, const F &is_less) {
```
For consistency, could you please mark all these functions as `LIBC_INLINE`?

Voultapher replied:
I'm surprised that template function should be marked with `inline`, AFAIK they can't run into ODR issues, or is this about giving the compiler an inlining hint?

SchrodingerZhu reviewed (approved):
LGTM with some minor comments.

nickdesaulniers commented on libc/src/stdlib/qsort_util.h:
```diff
@@ -27,11 +31,51 @@
+                   const F &is_less) {
 #if LIBC_QSORT_IMPL == LIBC_QSORT_QUICK_SORT
-constexpr auto sort = quick_sort;
+  unstable_sort_impl<true, F>(array, array_len, elem_size, is_less);
 #elif LIBC_QSORT_IMPL == LIBC_QSORT_HEAP_SORT
-constexpr auto sort = heap_sort;
+  unstable_sort_impl<false, F>(array, array_len, elem_size, is_less);
 #endif
```
I think this can be expressed more concisely.
```c++
// Should we use qsort or heap sort?
#define USE_QSORT (LIBC_QSORT_IMPL) == (LIBC_QSORT_QUICK_SORT)
unstable_sort_impl<USE_QSORT, F>(array, array_len, elem_size, is_less);
```

nickdesaulniers replied:
> but in future it may be useful to have others (e.g. mergesort)

ACK. timsort comes to mind.

Voultapher replied:
@michaelrj-google in my experience future proofing code for some vague future that is not concretely planned is a recipe for unnecessary complexity. Specifically here, if at some future point more than these two options will be added the `bool` is changed to an `enum` and problem solved. That said I doubt if adding a mergesort here is a good idea, since the current implementation is in-place and suddenly allocating large buffers in a function that previously didn't is recipe for unhappy users.

I think we need the impl function so we can test the code users will use, more on that in the other test comment.

michaelrj-google reviewed (changes requested):
There are some changes I'd like to make before this gets merged. Typing them up now.

michaelrj-google commented on libc/test/src/stdlib/SortingTest.h:
```diff
@@ -315,15 +278,76 @@ class SortingTest : public LIBC_NAMESPACE::testing::Test {
+    // incorrect association between alignment and element size.
+    alignas(1) uint8_t buf[BUF_SIZE];
+
+    const auto fill_buf = [&buf](size_t elem_size) {
+      for (size_t i = 0; i < BUF_SIZE; ++i) {
+        buf[i] = 0;
+      }
+
+      for (size_t elem_i = 0, buf_i = 0; elem_i < ARRAY_LEN; ++elem_i) {
+        const uint8_t elem_val = ARRAY_INITIAL_VALS[elem_i];
+        for (size_t elem_byte_i = 0; elem_byte_i < elem_size; ++elem_byte_i) {
+          buf[buf_i] = elem_val;
+          buf_i += 1;
+        }
+      }
+    };
```
this might be better as a standalone function instead of a lambda.

Voultapher replied:
A standalone function is useful if we expect re-use, I don't see this being used anywhere but in that function. Plus making standalone would mean moving it to the top of the file which makes understanding the test needlessly complicated. This way domain specific logic is in one place. It's not like we have a 100 LoC function here.

michaelrj-google commented on libc/src/stdlib/qsort_util.h:
```diff
@@ -15,9 +15,13 @@
 #define LIBC_QSORT_QUICK_SORT 1
 #define LIBC_QSORT_HEAP_SORT 2
 
+#ifdef LIBC_OPTIMIZE_FOR_SIZE
+#define LIBC_QSORT_IMPL LIBC_QSORT_HEAP_SORT
+#else
```
There doesn't need to be a specific check here. The algorithm selection is done in [`config.json`](https://github.com/llvm/llvm-project/blob/main/libc/config/config.json#L96)

Voultapher replied:
I need more information here, my idea was to keep the existing `LIBC_CONF_QSORT_IMPL` flag for compatibility reasons and add a new one that could be re-used across different areas of the library. It's not immediately clear to me how to do that with `config.json` and what the implications would be, and even then it would have to be used here in this way also, no?

Voultapher replied:
I agree let's pull it out.

michaelrj-google reviewed (commented):
This is getting close to complete, I think after one more round of review it'll be good to go. I'm hopeful we can get this landed before I go on break for the holidays.

orlp commented on libc/src/stdlib/qsort_r.cpp:
```diff
@@ -19,13 +19,11 @@ LLVM_LIBC_FUNCTION(void, qsort_r,
-                             elem_size, c);
 
-  internal::sort(arr);
+  internal::unstable_sort(array, array_size, elem_size,
+                          [compare, arg](const void *a, const void *b) -> bool {
+                            return compare(a, b, arg) < 0;
+                          });
```
@michaelrj-google I know this is ultimately subjective so I can't do better than an appeal to authority, but... as someone that knows a thing or two about sorting, the last decade of sorting theory has firmly discarded the ternary comparison in favor of a binary one. The public interface (obviously) can't be changed to match this, but the implementation however can be modern. I would strongly recommend to use the `is_less` paradigm in the implementation for consistency with all other modern sorting implementations, especially because the implemented sorting algorithm does not use the ternary nature of the comparison function at all.

michaelrj-google replied:
My concern is mainly around readability. I found it hard to determine what `is_less` did in the primary qsort files since it's always passed with a templated type, and I had a hard time finding where it came from because it's never declared as `is_less`, it's just an anonymous lambda. I don't think we can easily fix the template issue, since the lambdas for qsort and qsort_r have different types, but we might be able to solve the lack of a declaration.
```suggestion
  const auto is_less = [compare, arg](const void *a, const void *b) -> bool {
    return compare(a, b, arg) < 0;
  };
  internal::unstable_sort(array, array_size, elem_size, is_less);
```
This would make it much easier to find the source of `is_less`, without materially changing the code. Does this seem like a good solution to you?

Voultapher replied:
I appreciate your concern about readability and like the idea of pulling out the comparison function to a named variable, will do.

michaelrj-google commented on libc/test/src/stdlib/SortingTest.h:
```diff
@@ -7,19 +7,21 @@
+#include "src/stdlib/qsort.h"
 #include "test/UnitTest/Test.h"
 
+#include <iostream>
+
```
We can't include libc++ headers, so this needs to be removed.

Voultapher replied:
Good catch, that's debug leftover.

Voultapher commented on libc/test/src/stdlib/heap_sort_test.cpp:
```diff
@@ -7,10 +7,18 @@
+void heap_sort(void *array, size_t array_size, size_t elem_size,
+               int (*compare)(const void *, const void *)) {
+
+  constexpr bool USE_QUICKSORT = false;
+
+  LIBC_NAMESPACE::internal::unstable_sort_impl<USE_QUICKSORT>(
+      array, array_size, elem_size,
+      [compare](const void *a, const void *b) noexcept -> bool {
+        return compare(a, b) < 0;
+      });
```
Well I guess it depends on *what* we want to test. Do we want to test that `qsort` users get the behavior they want, or do we want to unit test an implementation detail. Personally I find the first much more important than the latter. By limiting the logic difference between a call to `qsort` and this test function we minimize the chance of missing bugs in the current and future logic contained in `unstable_sort_impl`.

Voultapher commented on libc/test/src/stdlib/quick_sort_test.cpp:
```diff
@@ -7,10 +7,17 @@
-  LIBC_NAMESPACE::internal::quick_sort(array);
+void quick_sort(void *array, size_t array_size, size_t elem_size,
+                int (*compare)(const void *, const void *)) {
+  constexpr bool USE_QUICKSORT = true;
+
+  LIBC_NAMESPACE::internal::unstable_sort_impl<USE_QUICKSORT>(
+      array, array_size, elem_size,
+      [compare](const void *a, const void *b) noexcept -> bool {
+        return compare(a, b) < 0;
+      });
```
Same argument as https://github.com/llvm/llvm-project/pull/120450#discussion_r1893719735 but arguably even more important since if we would call `quick_sort` directly we wouldn't test the Lomuto partition.

Voultapher committed:
Apply review comments

michaelrj-google reviewed (approved):
Looks pretty much done, I've left one last comment but it should be fine to land after addressing that.

Voultapher committed:
Apply review comments

Voultapher commented:
I think this should be it now, no more unresolved topics as far as I can tell.

SchrodingerZhu commented:
merge according to previous approval. Thank you again for the patch!

SchrodingerZhu commented:
@Voultapher can u make a quick fix to the build error? If you are not available right now, I will need to revert the patch and reland this with your fix.

SchrodingerZhu commented:
I will go ahead and revert this to keep the bots happy.
Looking forward to your new patch with fix.

Voultapher commented:
@michaelrj-google from what I know Google has extensive performance monitoring in production, once this change reaches systems it would be interesting to get some feedback if it actually improves real world cases. A lot of the perf critical sorting is done via vqsort as far as I understand, but this could still have some impact would be great to know - with the appropriate lack of details of course.

SchrodingerZhu commented:
@Voultapher In case you are interested:

I did not mention too much details about __builtin_memcpy in the before since both functions work as expected.

Basically, in GNU extension, many libc function has its builtin counterparts. They are used to do libcall optimizations. So semantically, calling __builtin_XXX should be understood as the same as calling the c function ::XXX. However, due to lib all optimizations, all of them are usually transformed into compiler intrinsics. For trivial case such as fixed size copy, they are optimized into direct hardware instructions. For large sizes or the cases where it is hard to decide the size or the most efficient copy algorithm, it falls back to a call to memcpy.

__builtin_inline_memcpy is a clang-specific intrinsic. It supports only static sizes and it does not have the same semantic of ::memcpy. It guarantees that no libcall will be ever emitted. Hence it is a good candidate if one wants to use builtin intrinsics to implement memcpy itself.
The LLVM compiler will decides the most proper hardware instructions to do the copy according to the cost model . E.g. on x86, small sized objects are copied using mov or sse/avx2/avx512. And larger sizes are copied using rep movsb if erms is available or just loops on older machines.

Voultapher commented:
@SchrodingerZhu thanks for the explanation.

michaelrj-google commented:
@Voultapher I'll be sure to keep you up to date when we roll this out (though that may take a while, the testing process is slow). It will be replacing an older version of glibc's qsort so I'm optimistic there will be decent performance gains. Anyone internally who is specifically using another implementation (like vqsort) won't be affected, since we're just replacing the C symbol.
